Used this tutorial: https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html

code can be found here: https://gist.github.com/fchollet/0830affa1f7f19fd47b06d4cf89ed44d

In [6]:
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D
from keras.layers import Activation, Dropout, Flatten, Dense
from keras import backend as K


# dimensions of our images.
img_width, img_height = 150, 150

train_data_dir = 'data/train'
validation_data_dir = 'data/validation'
nb_train_samples = 2000
nb_validation_samples = 800
epochs = 50
batch_size = 16

if K.image_data_format() == 'channels_first':
    input_shape = (3, img_width, img_height)
else:
    input_shape = (img_width, img_height, 3)

model = Sequential()
model.add(Conv2D(32, (3, 3), input_shape=input_shape))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Flatten())
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

train_datagen = ImageDataGenerator(
    rescale=1. / 255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True)

# this is the augmentation configuration we will use for testing:
# only rescaling
test_datagen = ImageDataGenerator(rescale=1. / 255)

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
    validation_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='binary')

model.fit_generator(
    train_generator,
    steps_per_epoch=nb_train_samples // batch_size,
    epochs=epochs,
    validation_data=validation_generator,
    validation_steps=nb_validation_samples // batch_size)


Found 1324 images belonging to 2 classes.
Found 123 images belonging to 2 classes.
Instructions for updating:
Use tf.cast instead.
Epoch 1/50
125/125 [==============================] - 43s 340ms/step - loss: 0.4698 - accuracy: 0.7981 - val_loss: 0.4273 - val_accuracy: 0.7558
Epoch 2/50
125/125 [==============================] - 41s 328ms/step - loss: 0.2824 - accuracy: 0.8956 - val_loss: 0.0882 - val_accuracy: 0.9610
Epoch 3/50
125/125 [==============================] - 40s 318ms/step - loss: 0.2169 - accuracy: 0.9218 - val_loss: 0.1532 - val_accuracy: 0.9584
Epoch 4/50
125/125 [==============================] - 39s 316ms/step - loss: 0.2027 - accuracy: 0.9232 - val_loss: 0.0178 - val_accuracy: 0.9438
Epoch 5/50
125/125 [==============================] - 39s 316ms/step - loss: 0.1655 - accuracy: 0.9378 - val_loss: 0.2372 - val_accuracy: 0.9584
Epoch 6/50
125/125 [==============================] - 566s 5s/step - loss: 0.1590 - accuracy: 0.9464 - val_loss: 0.4186 - val_accuracy: 0.9649
E

In [7]:
model.save_weights('bed_table_classifier.h5')

## We achieved 96% accuracy on the validation data!
We can try to tune the hyperparameters to improve this still. Additionally, we can see if there are any data augmentation or preprocessing we can do to improve this score